plugins/chain_plugin/trx_finality_status_processing.cpp

```diff
@@ -1,12 +1,215 @@
+      // need to verify not in a block since the index will transition from old failed transactions to newer successful transactions
+      for (iter = indx.lower_bound(boost::make_tuple(chain::block_id_type{}, failure_expiry));
+           iter != indx.end() && iter->block_id == chain::block_id_type{};
+           ++iter) {
+         remove_trxs.push_back(iter->trx_id);
+      }
+
+      for (const auto& trx_id : remove_trxs) {
+         _storage.erase(trx_id);
+      }
+      return !remove_trxs.empty();
+   }
+
+   void trx_finality_status_processing_impl::ensure_storage(const fc::time_point& now) {
+#warning TODO logging
+      const int64_t remaining_storage = _storage.memory_size() - _max_storage;
+      if (remaining_storage > 0) {
+         return;
+      }
+
+      auto percentage = [](uint64_t mem) {
+         const uint64_t pcnt = 90;
+         return (mem * pcnt)/100;
+      };
+      // determine how much we need to free to get back to at least the desired percentage of the storage
+      int64_t storage_to_free = _storage.memory_size() - percentage(_max_storage) - remaining_storage;
+      const auto& indx = _storage.index().get<by_success_status_expiry>();
+      std::deque<chain::transaction_id_type> remove_trxs;
+
+      const auto failure_rev_end = indx.rend();
+
+      auto success_rev_iter = indx.rbegin();
+      // find the start (or reversed last) of the valid block number range
+      auto failure_rev_iter = boost::make_reverse_iterator(indx.upper_bound(boost::make_tuple(chain::block_id_type{}, fc::time_point{})));
+      // rev iterate to make it the end (instead of last)
+      if (failure_rev_iter != failure_rev_end) {
+         ++failure_rev_iter;
+      }
+
+      const auto success_rev_end = failure_rev_iter;
+
+      const double now_double = now.time_since_epoch().count();
+      auto percent_of_time = [now_double](const fc::time_point& time, const  fc::microseconds& duration) {
+         return (now_double - time.time_since_epoch().count())/duration.count();
+      };
+      const double ignore = 1.01; // ignored because it is greater than 100%
+      uint32_t earliest_block = finality_status::no_block_num;
+      while (storage_to_free > 0 && (success_rev_iter != success_rev_end || failure_rev_iter != failure_rev_end)) {
+         const auto success = success_rev_iter != success_rev_end ? percent_of_time(success_rev_iter->received, _success_duration) : ignore;
+         const auto failure = failure_rev_iter != failure_rev_end ? percent_of_time(failure_rev_iter->received, _failure_duration) : ignore;
+         if (success < failure) {
+            storage_to_free -= success_rev_iter->size();
+            const auto block_num = chain::block_header::num_from_id(success_rev_iter->block_id);
+            if (earliest_block == finality_status::no_block_num || block_num < earliest_block ) {
+               earliest_block = block_num;
+            }
+            remove_trxs.push_back(success_rev_iter++->trx_id);
```

Comment:
I find ++-> rather hard to read. I would prefer:
```
remove_trxs.push_back(success_rev_iter->trx_id);
++success_rev_iter;
```

Reply:
OBE